subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_espi.c

```diff
@@ -0,0 +1,135 @@
+	const uint16_t rx_valid_data_size = rx_header->data_len + RX_HEADER_SIZE;
+
+	if (event_type != ESPI_PERIPHERAL_EC_HOST_CMD) {
+		return;
+	}
+
+	/* Only support version 3 and make sure the number of bytes to copy is not
+	 * bigger than rx buf size or the shared memory size
+	 */
+	if (rx_header->prtcl_ver != 3 ||
+	    rx_valid_data_size > CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER ||
+	    rx_valid_data_size > shared_size) {
+		memcpy(hc_espi->rx_ctx->buf, (void *)rx_header, RX_HEADER_SIZE);
+		hc_espi->rx_ctx->len = RX_HEADER_SIZE;
+	} else {
+		memcpy(hc_espi->rx_ctx->buf, (void *)rx_header, rx_valid_data_size);
+		hc_espi->rx_ctx->len = rx_valid_data_size;
+	}
+
+	/* Even in case of errors, let the general handler send response */
+	k_sem_give(&hc_espi->rx_ctx->handler_owns);
+}
+
+static int ec_host_cmd_espi_init(const struct ec_host_cmd_backend *backend, const void *config,
+				 struct ec_host_cmd_rx_ctx *rx_ctx, struct ec_host_cmd_tx_buf *tx)
+{
+	struct ec_host_cmd_espi *hc_espi = (struct ec_host_cmd_espi *)backend->ctx;
+
+	hc_espi->espi_dev = ((const struct ec_host_cmd_espi_config *)config)->dev;
+
+	if (!device_is_ready(hc_espi->espi_dev)) {
+		return -ENODEV;
+	}
+
+	/* eSPI backend needs rx buffer provided by the handler */
+	if (!rx_ctx->buf) {
+		return -EIO;
+	}
+
+	/* eSPI backend uses the shared memory as tx buffer */
+	if (tx->buf) {
+		LOG_WRN("Providing unused buffer");
+	}
+
+	hc_espi->rx_ctx = rx_ctx;
+	hc_espi->tx = tx;
+
+	espi_init_callback(&hc_espi->espi_cb, espi_handler, ESPI_BUS_PERIPHERAL_NOTIFICATION);
+	espi_add_callback(hc_espi->espi_dev, &hc_espi->espi_cb);
+	/* Use shared memory as the tx buffer */
+	espi_read_lpc_request(hc_espi->espi_dev, ECUSTOM_HOST_CMD_GET_PARAM_MEMORY,
+			      (uint32_t *)&tx->buf);
+	espi_read_lpc_request(hc_espi->espi_dev, ECUSTOM_HOST_CMD_GET_PARAM_MEMORY_SIZE,
+			      &tx->len_max);
+
+	return 0;
+}
+
+static int ec_host_cmd_espi_send(const struct ec_host_cmd_backend *backend)
+{
+	struct ec_host_cmd_espi *hc_espi = (struct ec_host_cmd_espi *)backend->ctx;
+	struct ec_host_cmd_response_header *resp_hdr = hc_espi->tx->buf;
+	uint32_t result = resp_hdr->result;
+
+	/* Data to transfer are already in the tx buffer (shared memory) */
+	return espi_write_lpc_request(hc_espi->espi_dev, ECUSTOM_HOST_CMD_SEND_RESULT, &result);
+}
+
+static const struct ec_host_cmd_backend_api ec_host_cmd_api = {
+	.init = &ec_host_cmd_espi_init,
+	.send = &ec_host_cmd_espi_send,
+};
+
+EC_HOST_CMD_ESPI_DEFINE(ec_host_cmd_espi);
+struct ec_host_cmd_backend *ec_host_cmd_backend_get_espi(void)
+{
+	return &ec_host_cmd_espi;
+}
+
+#if DT_NODE_EXISTS(DT_CHOSEN(zephyr_host_cmd_backend))
+static int host_cmd_init(const struct device *arg)
+{
+	ARG_UNUSED(arg);
+	const struct device *const dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_host_cmd_backend));
+	struct ec_host_cmd_espi_config config = {.dev = dev};
+
+	ec_host_cmd_init(ec_host_cmd_backend_get_espi(), (void *)&config);
```

Comment:
If I understand correctly, if you do not set a specific backend as chosen, then you have to manually call the init function with manually selected backend. I think that this custom code should have a pointers to available devices, probe them or do other magic and then call the ec_host_cmd_init using pointer to the device. Generic API shouldn't have any knowledge about the drivers implementation.

Reply:
First of all, it is not a driver :-) a user(application code) doesn't use the host command API. The user only has to set the `DT_CHOSEN(zephyr_host_cmd_backend)` node, or call the `ec_host_cmd_init` function manually. `ec_host_cmd_backend_get_*`  functions are created for that purpose. The shell subsystem has pretty similar mechanism.
Anyway, I added comments about `DT_CHOSEN(zephyr_host_cmd_backend)`  to header files.